adb/adb_debug.py

```diff
@@ -36,34 +36,173 @@
     rsa_signer = None
 
 
-gflags.ADOPT_module_key_flags(common_cli)
+def Devices(args):
+  """Lists the available devices.
+
+  Mimics 'adb devices' output:
+    List of devices attached
+    015DB7591102001A        device        1,2
+  """
+  for d in adb_commands.AdbCommands.Devices():
+    if args.output_port_path:
+      print('%s\tdevice\t%s' % (
+            d.serial_number, ','.join(str(p) for p in d.port_path)))
+    else:
+      print('%s\tdevice' % d.serial_number)
+  return 0
+
+
+def List(self, *args):
+  """Prints a directory listing."""
+  files = adb_commands.AdbCommands.List(self, args[0])
+  files.sort(key=lambda x: x.filename)
+  maxname = max(len(f.filename) for f in files)
+  maxsize = max(len(str(f.size)) for f in files)
+  for f in files:
+    mode = (
+      ('d' if stat.S_ISDIR(f.mode) else '-') +
+      ('r' if f.mode & stat.S_IRUSR else '-') +
+      ('w' if f.mode & stat.S_IWUSR else '-') +
+      ('x' if f.mode & stat.S_IXUSR else '-') +
+      ('r' if f.mode & stat.S_IRGRP else '-') +
+      ('w' if f.mode & stat.S_IWGRP else '-') +
+      ('x' if f.mode & stat.S_IXGRP else '-') +
+      ('r' if f.mode & stat.S_IROTH else '-') +
+      ('w' if f.mode & stat.S_IWOTH else '-') +
+      ('x' if f.mode & stat.S_IXOTH else '-'))
+    t = time.gmtime(f.mtime)
+    yield '%s %*d %04d-%02d-%02d %02d:%02d:%02d %-*s\n' % (
+        mode, maxsize, f.size,
+        t.tm_year, t.tm_mon, t.tm_mday, t.tm_hour, t.tm_min, t.tm_sec,
+        maxname, f.filename)
+
+
+def Logcat(self, *args):
+  """Shortcut to \'shell logcat\'."""
+  return adb_commands.AdbCommands.StreamingShell(
+      self, 'logcat ' + ' '.join(args))
+
+
+def Shell(self, *args):
+  """Run a command on the device"""
+  return adb_commands.AdbCommands.StreamingShell(self, ' '.join(args))
+
+
+def main():
+  common = common_cli.GetCommonArguments()
+  common.add_argument(
+      '--rsa_key_path', action='append', default=[],
+      metavar='~/.android/adbkey',
+      help='RSA key(s) to use, use multiple times to load mulitple keys')
+  common.add_argument(
+    '--auth_timeout_s', default=60., metavar='60', type=int,
+    help='Seconds to wait for the dialog to be accepted when using '
+         'authenticated ADB.')
+  device = common_cli.GetDeviceArguments()
+  parents = [common, device]
+
+  parser = argparse.ArgumentParser(
+      description=sys.modules[__name__].__doc__, parents=[common])
+  subparsers = parser.add_subparsers(title='Commands', dest='command_name')
```

Comment:
This all makes me so sad! If there were less boilerplate (maybe the positional=[] arg is enough?) then it wouldn't be so bad. At the moment, a hundred lines of argument parsing is an order-of-magnitude too much, sorry